src/uri.cr

```diff
@@ -265,8 +375,15 @@ class URI
         return i
       end
 
-      io.write_byte (first_num * 16 + second_num).to_u8
+      encoded = (first_num * 16 + second_num).to_u8
       i += 1
+      if yield encoded
```

Comment:
According to RFC 3986, URI consists of reserved characters (e.g. separator symbols like `':'`, `'/'`, `'?'` and some symbols like `'('`, `')'`) and unreserved characters (e.g. alphabets, digits and some symbols like `'_'`, `'-'`). This characters set dose not contains multi-byte characters, so multi-byte characters are being percent-encoded always, and unreserved characters are not being percent-encoded always (See RFC 3986, [§2.3.](https://tools.ietf.org/html/rfc3986#section-2.3) and [§2.4.](https://tools.ietf.org/html/rfc3986#section-2.4))).
But, reserved characters, they cannot decide to be being encoded without context. For example, you have the string `"/hello%3Fworld=foo"`, `"%3F"` means `"?"`. If the string is URI component (path), it is decoded to `"/hello?world=foo"`, however it produces broken URI if the string is full URI. In fact, JavaScript's `decodeURI("/hello%3Fworld=foo")` returns `"/hello%3Fworld=foo"`, and `decodeURIComponent("/hello%3Fworld=foo")` returns `"/hello?world=foo"`.

Summarizing the above, `URI#unescape`'s block decides whether to decode given reserved character and all reserved character is just one byte, so the block wants byte value and returns boolean value.

Another point, `URI#unescape_one` and `URI#unescape(io : IO)` are not depending character encoding. It is awesome because Crystal supports any encodings now. I don't want to break it.

However, yielded bytes might have to be restricted to `< 0x80`. How do you think?